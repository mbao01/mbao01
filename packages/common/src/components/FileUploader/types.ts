import type { Dispatch, Ref, SetStateAction } from "react";
import type { DropzoneOptions, DropzoneState } from "react-dropzone";

export type FileUploaderProps = {
  value: File[] | null;
  reSelect?: boolean;
  onValueChange: (value: File[] | null) => void;
  dropzoneOptions: DropzoneOptions;
  orientation?: "horizontal" | "vertical";
} & React.HTMLAttributes<HTMLDivElement>;

export type FileUploaderInputProps = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  "className"
> & {
  classes?: Partial<{
    all: string;
    accepted: string;
    rejected: string;
    default: string;
  }>;
};

export type DirectionOptions = "rtl" | "ltr" | undefined;

export type FileUploaderContextType = {
  dropzoneState: DropzoneState;
  isLOF: boolean;
  isFileTooBig: boolean;
  removeFileFromSet: (index: number) => void;
  removeAll: () => void;
  activeIndex: number;
  setActiveIndex: Dispatch<SetStateAction<number>>;
  orientation: "horizontal" | "vertical";
  direction: DirectionOptions;
  hiddenInputRef: Ref<HTMLInputElement>;
};
