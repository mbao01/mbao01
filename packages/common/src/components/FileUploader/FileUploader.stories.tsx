import type { Meta, StoryObj } from "@storybook/react";
import { useState } from "react";
import { FileIcon } from "@radix-ui/react-icons";
import { FileUploader } from "./FileUploader";

const FileSvgDraw = () => {
  return (
    <>
      <svg
        className="w-8 h-8 mb-3 text-base-content"
        aria-hidden="true"
        xmlns="http://www.w3.org/2000/svg"
        fill="none"
        viewBox="0 0 20 16"
      >
        <path
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          d="M13 13h3a3 3 0 0 0 0-6h-.025A5.56 5.56 0 0 0 16 6.5 5.5 5.5 0 0 0 5.207 5.021C5.137 5.017 5.071 5 5 5a4 4 0 0 0 0 8h2.167M10 15V6m0 0L8 8m2-2 2 2"
        />
      </svg>
      <p className="mb-1 text-sm text-base-content px-4">
        <span className="font-semibold">Click to upload</span>
        &nbsp; or drag and drop
      </p>
      <p className="text-xs text-base-content">SVG, PNG, JPG or GIF</p>
    </>
  );
};

const FileUpload = () => {
  const [files, setFiles] = useState<File[] | null>(null);

  const dropZoneConfig = {
    maxFiles: 5,
    maxSize: 1024 * 1024 * 4,
    multiple: true,
  };

  return (
    <FileUploader
      value={files}
      onValueChange={setFiles}
      dropzoneOptions={dropZoneConfig}
      className="relative bg-background rounded-lg p-2"
    >
      <FileUploader.Input className="outline-dashed outline-1 outline-neutral-content">
        <div className="flex items-center justify-center flex-col pt-3 pb-4 w-full ">
          <FileSvgDraw />
        </div>
      </FileUploader.Input>
      <FileUploader.Content>
        {files &&
          files.length > 0 &&
          files.map((file, i) => (
            <FileUploader.Item key={i} index={i}>
              <FileIcon className="h-4 w-4" />
              <span>{file.name}</span>
            </FileUploader.Item>
          ))}
      </FileUploader.Content>
    </FileUploader>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: FileUpload,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof FileUpload>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {};
