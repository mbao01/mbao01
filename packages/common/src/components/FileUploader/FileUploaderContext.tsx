import { createContext } from "react";
import { type FileUploaderContextType } from "./types";

export const FileUploaderContext =
  createContext<FileUploaderContextType | null>(null);
