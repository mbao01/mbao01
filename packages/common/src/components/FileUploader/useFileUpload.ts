import { useContext } from "react";
import { FileUploaderContext } from "./FileUploaderContext";

export const useFileUpload = () => {
  const context = useContext(FileUploaderContext);
  if (!context) {
    throw new Error("useFileUpload must be used within a FileUploaderProvider");
  }
  return context;
};
