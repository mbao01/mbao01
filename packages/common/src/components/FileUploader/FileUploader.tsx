"use client";

import { forwardRef, useCallback, useEffect, useRef, useState } from "react";
import { useDropzone, FileRejection } from "react-dropzone";
import { toast } from "sonner";
import { TrashIcon } from "@radix-ui/react-icons";
import type {
  DirectionOptions,
  FileUploaderInputProps,
  FileUploaderProps,
} from "./types";
import { cn } from "../../utilities";
import { FileUploaderContext } from "./FileUploaderContext";
import { useFileUpload } from "./useFileUpload";

export const FileUploader = ({
  className,
  dropzoneOptions,
  value,
  onValueChange,
  reSelect,
  orientation = "vertical",
  children,
  dir,
  ...props
}: FileUploaderProps) => {
  const hiddenInputRef = useRef<HTMLInputElement>(null);
  const [isFileTooBig, setIsFileTooBig] = useState(false);
  const [isLOF, setIsLOF] = useState(false);
  const [activeIndex, setActiveIndex] = useState(-1);
  const {
    accept = {
      "image/*": [".jpg", ".jpeg", ".png", ".gif"],
    },
    maxFiles = 1,
    maxSize = 4 * 1024 * 1024,
    multiple = true,
  } = dropzoneOptions;

  const reSelectAll = maxFiles === 1 ? true : reSelect;
  const direction: DirectionOptions = dir === "rtl" ? "rtl" : "ltr";

  const removeFileFromSet = useCallback(
    (i: number) => {
      if (!value) return;
      const newFiles = value.filter((_, index) => index !== i);
      onValueChange(newFiles);
    },
    [value, onValueChange]
  );

  const removeAll = useCallback(() => {
    onValueChange(null);
  }, [onValueChange]);

  const handleKeyDown = useCallback(
    (e: React.KeyboardEvent<HTMLDivElement>) => {
      const files = value ?? [];
      const KEY_LIST = [
        "ArrowLeft",
        "ArrowRight",
        "ArrowUp",
        "ArrowDown",
        "Enter",
        "Space",
        "Delete",
        "Backspace",
        "Escape",
      ];

      if (KEY_LIST.includes(e.key)) {
        e.preventDefault();
        e.stopPropagation();

        const moveNext = () => {
          const nextIndex = activeIndex + 1;
          setActiveIndex(nextIndex > files.length - 1 ? 0 : nextIndex);
        };

        const movePrev = () => {
          const nextIndex = activeIndex - 1;
          setActiveIndex(nextIndex < 0 ? files.length - 1 : nextIndex);
        };

        const prevKey =
          orientation === "horizontal"
            ? direction === "ltr"
              ? "ArrowLeft"
              : "ArrowRight"
            : "ArrowUp";

        const nextKey =
          orientation === "horizontal"
            ? direction === "ltr"
              ? "ArrowRight"
              : "ArrowLeft"
            : "ArrowDown";

        if (e.key === nextKey) {
          moveNext();
        } else if (e.key === prevKey) {
          movePrev();
        } else if (e.key === "Enter" || e.key === "Space") {
          if (activeIndex === -1) {
            dropzoneState.inputRef.current?.click();
          }
        } else if (e.key === "Delete" || e.key === "Backspace") {
          if (activeIndex !== -1) {
            removeFileFromSet(activeIndex);
            if (files.length - 1 === 0) {
              setActiveIndex(-1);
              return;
            }
            movePrev();
          }
        } else if (e.key === "Escape") {
          setActiveIndex(-1);
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [value, activeIndex, removeFileFromSet]
  );

  const onDrop = useCallback(
    (acceptedFiles: File[], rejectedFiles: FileRejection[]) => {
      const files = acceptedFiles;

      if (!files) {
        toast.error("file error , probably too big");
        return;
      }

      const newValues: File[] = value ? [...value] : [];

      if (reSelectAll) {
        newValues.splice(0, newValues.length);
      }

      files.forEach((file) => {
        if (newValues.length < maxFiles) {
          newValues.push(file);
        }
      });

      onValueChange(newValues);

      if (hiddenInputRef.current) {
        // Note the specific way we need to munge the file into the hidden input
        // https://stackoverflow.com/a/68182158/1068446
        const dataTransfer = new DataTransfer();
        newValues.forEach((v) => {
          dataTransfer.items.add(v);
        });
        hiddenInputRef.current.files = dataTransfer.files;
      }

      if (rejectedFiles.length > 0) {
        for (const rejectedFile of rejectedFiles) {
          if (rejectedFile.errors[0]?.code === "file-too-large") {
            toast.error(
              `File is too large. Max size is ${maxSize / 1024 / 1024}MB`
            );
            break;
          }
          if (rejectedFile.errors[0]?.message) {
            toast.error(rejectedFile.errors[0].message);
            break;
          }
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [reSelectAll, value]
  );

  useEffect(() => {
    if (!value) return;
    if (value.length === maxFiles) {
      setIsLOF(true);
      return;
    }
    setIsLOF(false);
  }, [value, maxFiles]);

  const opts = dropzoneOptions
    ? dropzoneOptions
    : { accept, maxFiles, maxSize, multiple };

  const dropzoneState = useDropzone({
    ...opts,
    onDrop,
    onDropRejected: () => setIsFileTooBig(true),
    onDropAccepted: () => setIsFileTooBig(false),
  });

  return (
    <FileUploaderContext.Provider
      value={{
        dropzoneState,
        isLOF,
        isFileTooBig,
        removeFileFromSet,
        removeAll,
        activeIndex,
        setActiveIndex,
        orientation,
        direction,
        hiddenInputRef,
      }}
    >
      <div
        onKeyDownCapture={handleKeyDown}
        className={cn("grid w-full overflow-hidden", className, {
          "gap-2": value && value.length > 0,
        })}
        dir={dir}
        {...props}
        tabIndex={-1}
      >
        {children}
      </div>
    </FileUploaderContext.Provider>
  );
};

FileUploader.displayName = "FileUploader";

const FileUploaderContent = forwardRef<
  HTMLDivElement,
  React.HTMLAttributes<HTMLDivElement>
>(({ children, className, ...props }, ref) => {
  const { orientation } = useFileUpload();
  const containerRef = useRef<HTMLDivElement>(null);

  return (
    <div
      className={cn("w-full px-1")}
      ref={containerRef}
      aria-description="content file holder"
    >
      <div
        {...props}
        ref={ref}
        className={cn(
          "flex rounded-xl gap-1",
          orientation === "horizontal" ? "flex-raw flex-wrap" : "flex-col",
          className
        )}
      >
        {children}
      </div>
    </div>
  );
});

FileUploaderContent.displayName = "FileUploaderContent";

const FileUploaderItem = forwardRef<
  HTMLDivElement,
  { index: number } & React.HTMLAttributes<HTMLDivElement>
>(({ className, index, children, ...props }, ref) => {
  const { removeFileFromSet, activeIndex, direction } = useFileUpload();
  const isSelected = index === activeIndex;
  return (
    <div
      ref={ref}
      className={cn(
        "h-6 p-1 justify-between cursor-pointer relative rounded",
        { "bg-base-300": isSelected },
        className
      )}
      {...props}
    >
      <div className="text-sm leading-none tracking-tight flex items-center gap-1.5 h-full w-full">
        {children}
      </div>
      <button
        type="button"
        className={cn(
          "absolute",
          direction === "rtl" ? "top-1 left-1" : "top-1 right-1"
        )}
        onClick={() => removeFileFromSet(index)}
      >
        <span className="sr-only">remove item {index}</span>
        <TrashIcon
          className={cn(
            "w-4 h-4 shrink-0 hover:stroke-destructive duration-200 ease-in-out",
            { "text-error": isSelected }
          )}
        />
      </button>
    </div>
  );
});

FileUploaderItem.displayName = "FileUploaderItem";

const FileUploaderInput = ({
  classes,
  children,
  ...props
}: FileUploaderInputProps) => {
  const { dropzoneState, isFileTooBig, isLOF, hiddenInputRef } =
    useFileUpload();
  const rootProps = isLOF ? {} : dropzoneState.getRootProps();
  return (
    <div {...props} className="relative w-full">
      <div
        className={cn(
          classes?.all,
          "w-full rounded-lg duration-300 ease-in-out",
          isLOF ? "opacity-50 cursor-not-allowed" : "cursor-pointer",
          dropzoneState.isDragAccept
            ? classes?.accepted
            : dropzoneState.isDragReject || isFileTooBig
              ? classes?.rejected
              : classes?.default
        )}
        {...rootProps}
      >
        {children}
      </div>
      <input
        ref={hiddenInputRef}
        {...props}
        type="file"
        tabIndex={-1}
        className="hidden opacity-0"
      />
      <input
        ref={dropzoneState.inputRef}
        disabled={isLOF}
        {...dropzoneState.getInputProps()}
        tabIndex={-1}
        className={cn({ "cursor-not-allowed": isLOF })}
      />
    </div>
  );
};

FileUploaderInput.displayName = "FileInput";

FileUploader.Content = FileUploaderContent;
FileUploader.Input = FileUploaderInput;
FileUploader.Item = FileUploaderItem;
