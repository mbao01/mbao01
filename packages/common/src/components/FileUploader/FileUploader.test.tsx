import { useState } from "react";
import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { FileUploader } from "./FileUploader";

describe("FileUploader", () => {
  const FileUpload = () => {
    const [files, setFiles] = useState<File[] | null>(null);

    const dropZoneConfig = {
      maxFiles: 5,
      maxSize: 1024 * 1024 * 4,
      multiple: true,
    };

    return (
      <FileUploader
        value={files}
        onValueChange={setFiles}
        dropzoneOptions={dropZoneConfig}
        className="relative bg-background rounded-lg p-2"
      >
        <FileUploader.Input
          classes={{ all: "outline-dashed outline-1 outline-neutral-content" }}
        >
          Click here to upload
        </FileUploader.Input>
      </FileUploader>
    );
  };

  it("should render", () => {
    const { asFragment } = render(<FileUpload />);

    expect(screen.getByText("Click here to upload")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
