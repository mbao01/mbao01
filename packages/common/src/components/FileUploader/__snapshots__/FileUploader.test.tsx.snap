// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileUploader > should render 1`] = `
<DocumentFragment>
  <div
    class="grid w-full focus:outline-none overflow-hidden relative bg-background rounded-lg p-2"
    tabindex="0"
  >
    <div
      class="relative w-full cursor-pointer "
    >
      <div
        class="w-full rounded-lg duration-300 ease-in-out border-gray-300 outline-dashed outline-1 outline-neutral-content"
        role="presentation"
        tabindex="0"
      >
        Click here to upload
      </div>
      <input
        style="opacity: 0; display: none;"
        type="file"
      />
      <input
        class=""
        multiple=""
        style="display: none;"
        tabindex="-1"
        type="file"
      />
    </div>
  </div>
</DocumentFragment>
`;
